### DISCUSSION TITLE
How to handle async adapters?

### DISCUSSION THREAD

**Discussion by s1n7ax** [nvim-neotest/neotest#322] [Q&A]
I'm working on https://github.com/nvim-java/nvim-java. So far, I have created only two APIs to run tests.

- `run_current_test_class`
- `debug_current_test_class`

Both these are ran using `nvim-dap` & test discovery is done by communicating to `jdtls` language server.

However, I would like to have an adapter so I can get all the neotest features come with it. However, I don't really understand how to do asynchronous stuff from the given adapter interface. I can see that you have the `@async` annotation in each method, however, there is no callback function passed to any APIs.

So I would like to get help on
- doing async stuff
For instance, there is a language sever command for checking if the given file is a test or not in jdtls. It's `java.project.isTestFile`. However calling this has to be done asynchronously. So my adapter function should look like this with a callback function. But looking at the interface there is not? So how should this be handled?

```lua
adapter.is_test_file = function(file_path, callback)
	jdtls.executeCommand({
		command = 'java.project.isTestFile',
		arguments = { uri },
	}, function(err, is_test_file)
		assert(err, 'could not check the file')

		callback(is_test_file)
	end)
end
```

- What should `build_spec` return? Is it a dap laucher config?

**Comment by rcarriga** [comment]
You can use the `nio` library that is bundled with neotest to wrap callback style functions 
```lua
local nio = require("nio")

local executeCommand = nio.wrap(jdtls.executeCommand, 2)

adapter.is_test_file = function(file_path)
  local uri = vim.uri_from_fname(file_path)
  local err, is_test_file = executeCommand({
    command = "java.project.isTestFile",
    arguments = { uri },
  })
  assert(err, "could not check the file")
  return is_test_file
end
```

**Comment by rcarriga** [comment]
> What should build_spec return? Is it a dap laucher config?

Here are the [interface](https://github.com/nvim-neotest/neotest/blob/39bb0f5fa9955c512c7e4a937ffc0a5d00207d34/lua/neotest/adapters/interface.lua#L33-L35) and [types](https://github.com/nvim-neotest/neotest/blob/39bb0f5fa9955c512c7e4a937ffc0a5d00207d34/lua/neotest/types/init.lua#L58-L64). For the dap strategy, you need to pass a `dap` table in the `context` which is a launcher config. [Here is what neotest-python returns](https://github.com/nvim-neotest/neotest-python/blob/27a2676aa02193d6ba4ad26238c09ededa49a9a7/lua/neotest-python/init.lua#L23-L33)

**Comment by s1n7ax** [comment]
@rcarriga nvim-java got a co-routine library https://github.com/nvim-java/lua-async-await. Seems like the one you are using and the one I'm using is not compatible.

I tried `request_sync` in `vim.lsp.client` too. It doesn't work either.

When I do this,

```lua
function neotest.Adapter.is_test_file(file_path)
  if
    file_path
    == "/home/s1n7ax/Workspace/demo/src/test/java/com/example/demo/DemoApplicationTests.java"
  then
    return true
  end

  return false
end
```
 
`discover_positions` function is being called for `DemoApplicationTests.java` file. 

```
filter_dir src/test/java/com/example/example
filter_dir src/main/java/com/example/example
filter_dir bin/test/com/example/demo/demo
filter_dir bin/main/com/example/demo/demo
filter_dir src/test/java/com/example/demo/demo
filter_dir src/main/java/com/example/demo/demo
discover_position/home/s1n7ax/Workspace/demo/src/test/java/com/example/demo/DemoApplicationTests.java
```

However, when following logic is used,

```lua
local function execute_command(cmd_info, timeout, buffer)
  timeout = timeout and timeout or 5000
  buffer = buffer and buffer or 0

  return jdtls().request_sync("workspace/executeCommand", cmd_info, timeout, buffer)
end
```

```lua
function neotest.Adapter.is_test_file(file_path)
  vim.print("is test file check: " .. file_path)

  local response = execute_command({
    command = "java.project.isTestFile",
    arguments = { vim.uri_from_fname(file_path) },
  })

  if response.err then
    error(response.err.message)
  end

  vim.print(response.result)
  return response.result
end
```

Output looks as follows. It's defiantly blocking and value is being returned but `discover_positions` never called.

```lua
filter_dir bin/test/com/example/demo/demo
filter_dir bin/main/com/example/demo/demo
filter_dir src/test/java/com/example/demo/demo
filter_dir src/main/java/com/example/demo/demo
is test file check: /home/s1n7ax/Workspace/demo/gradlew
is test file check: /home/s1n7ax/Workspace/demo/src/test/java/com/example/demo/DemoApplicationTests.java
true
is test file check: /home/s1n7ax/Workspace/demo/src/test/java/com/example/demo/DemoApplicationTests.java
true
is test file check: /home/s1n7ax/Workspace/demo/src/test/java/com/example/demo/DemoApplicationTests.java
true
```

**Reply by rcarriga** [reply]
I can't see how your async library would be interfering but since this is only partial code, it's impossible to rule out. From the snippets you've provided so far, there is no involvement of your async library right? (Not too sure what `jdtls` is)

Without a full reproduction example, there's not much I can do to help here unfortunately. You could try adding some log lines to the client file in neotest, and check the log. [Here](https://github.com/nvim-neotest/neotest/blob/39bb0f5fa9955c512c7e4a937ffc0a5d00207d34/lua/neotest/client/init.lua#L283-L295) is where the `is_test_file` function is used.

**Reply by s1n7ax** [reply]
No, I first tried to use my async library and yours both at the same time. It did not work. That's why I have tried above example which did not work either

**Comment by s1n7ax** [comment]
THIS IS THE implementation with my co-routine library. NOT the above example.

I pushed implementation here. This wont work without https://github.com/nvim-java-core dependency.
https://github.com/nvim-java/neotest-jdtls

I have this to call jdtls and get the is test status from server. I have made it a callback style so I can wrap it using your library
https://github.com/nvim-java/neotest-jdtls/blob/0fca10b753d16d33ff8b06b34cf9c4b0a257bd45/lua/neotest-jdtls/helper.lua?plain=1#L16-L28

Here is how I've wrapped and used it
https://github.com/nvim-java/neotest-jdtls/blob/0fca10b753d16d33ff8b06b34cf9c4b0a257bd45/lua/neotest-jdtls/init.lua?plain=1#L57-L67

Here is the output. As you can see from the logs, request is sent to get is test file check and response was received successfully.

```
is test: /home/s1n7ax/Workspace/demo/src/test/java/com/example/demo/DemoApplicationTests.java
*************************
{
  arguments = { "file:///home/s1n7ax/Workspace/demo/src/test/java/com/example/demo/DemoApplicationTests.java" },
  command = "java.project.isTestFile"
}
<function 1>
>>>>>>>>>>>>>>>>>>>>>>>>>
java.project.isTestFile success! response: 
true
res: 
true
Error executing vim.schedule lua callback: ...x/Workspace/nvim-java-core/lua/java-core/utils/async.lua:41: unhandled error ...n7ax/.local/share/nvim/lazy/nvim-nio/lua/nio/control.lua:133: Future already set
stack traceback:
	[C]: in function 'error'
	...x/Workspace/nvim-java-core/lua/java-core/utils/async.lua:41: in function 'callback'
	...nvim-java-core/lua/java-core/ls/clients/jdtls-client.lua:51: in function 'handler'
	...eovim-unwrapped-0.9.4/share/nvim/runtime/lua/vim/lsp.lua:1393: in function ''
	vim/_editor.lua: in function <vim/_editor.lua:0>
```
